File: audio_experiment/methods/overiva/overiva_mtlb_wrap.py
import importlib
import numpy as np
import overiva
# importlib.reload(overiva)

# X = np.asarray(X)

# print(X.shape)
# print(X.dtype)
# print(n_src)                                 
# print(n_iter)
# print(proj_back)
# print(W0.shape)
# print(model)
# print(init_eig)
# print(return_filters)

if W0 is not None:
    # print(W0.shape)
    if len(W0.shape)==2:
        W0 = np.expand_dims(W0,2);

if return_filters:
    [S, W] = overiva.overiva(np.array(X),n_src=n_src,n_iter=n_iter,proj_back=proj_back,model=model,init_eig=init_eig,return_filters=return_filters,W0=W0)
    W = np.array(W);
else:
    S      = overiva.overiva(np.array(X),n_src=n_src,n_iter=n_iter,proj_back=proj_back,model=model,init_eig=init_eig,return_filters=return_filters,W0=W0)
S = np.array(S);

File: audio_experiment/methods/overiva/overiva_mtlb_wrap.m
function [S, W] = overiva_mtlb_wrap(X, options)
    arguments
        X double,
        options.n_src (1,1) {mustBeInteger,mustBeNonnegative} = size(X,1), 
        options.n_iter (1,1) {mustBeInteger,mustBeNonnegative} = 20,
        options.proj_back (1,1) {mustBeNumericOrLogical} = false,
        options.W0 = py.None,
        options.model {mustBeMember(options.model,["laplace","gauss"])} = "laplace",
        options.init_eig (1,1) {mustBeNumericOrLogical} = false,
        options.return_filters (1,1) {mustBeNumericOrLogical} = true;
    end
    % X (n_chan,n_frames,n_freq)
    if (string(pyenv().Status))=="NotLoaded"
        pyenv('Version',"/home/jerry/anaconda3/envs/overiva_env/bin/python");
        py.numpy.array(1);
    end

    X_np = permute(X,[2,3,1]);
    X_np = py.numpy.array(X_np);
    
    if options.W0 ~= py.None
        W0 = permute(options.W0,[3,1,2]);
        W0 = py.numpy.array(W0);
    else
        W0 = options.W0;
    end

    [S, W] = pyrunfile("overiva_mtlb_wrap.py",["S","W"], X=X_np, ...
                                            n_src=int64(options.n_src), ...
                                            n_iter=int64(options.n_iter), ...
                                            proj_back=options.proj_back, ...
                                            W0=W0, ...
                                            model=options.model, ...
                                            init_eig=options.init_eig, ...
                                            return_filters=options.return_filters);
    
    S = double(S);
    if options.return_filters
        W = double(W);
        W = permute(W,[3,2,1]);
    end
    S = permute(S,[3,1,2]);

end

